### PR TITLE
[JS, CLI] Refactored runtime discovery in Genkit CLI.

### PR DESCRIPTION
**Pull request by apascal07**
New workflow:

```
alexpascal> genkit ui:start
Starting...

  Genkit Developer UI started at: http://localhost:4000
  To stop the UI, run `genkit ui:stop`.

alexpascal> genkit ui:stop
Stopping...

  Genkit Developer UI is stopped.
  To start the UI, run `genkit ui:start`.

alexpascal> genkit ui:start && npm start
Starting...

  Genkit Developer UI started at: http://localhost:4000
  To stop the UI, run `genkit ui:stop`.

> flow-sample1@1.0.0 start
> node lib/index.js

No flows registered in flow server.
Reflection server (12378) running on http://localhost:3100
Flow server running on http://localhost:3400
Set telemetry server URL to: http://localhost:4033
```

Checklist (if applicable):
- [x] Tested (manually, unit tested, etc.)
- [ ] Docs updated

### CODE REVIEW

File: genkit-tools/common/src/utils/utils.ts [hunk 1]
```diff
@@ -14,59 +14,119 @@
...
 
-interface PackageJson {
-  main: string;
+/**
+ * Finds the project root by looking for a `package.json` file.
+ */
+export async function findProjectRoot(): Promise<string> {
+  let currentDir = process.cwd();
+  while (currentDir !== path.parse(currentDir).root) {
+    const packageJsonPath = path.join(currentDir, 'package.json');
```
Comment (pavelgj) [inline genkit-tools/common/src/utils/utils.ts:28]: As a follow up, we might want to better handle CF3 case here. Package.json will be in the functions folder, but CWD might be at the root of firebase project folder, so this logic will fail to find the root.

File: genkit-tools/cli/src/commands/ui-start.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  findServersDir,
+  logger,
+  waitUntilHealthy,
+} from '@genkit-ai/tools-common/utils';
+import axios from 'axios';
+import { ChildProcess, spawn } from 'child_process';
+import * as clc from 'colorette';
+import { Command } from 'commander';
+import fs from 'fs/promises';
+import path from 'path';
```
Comment (pavelgj) [inline genkit-tools/cli/src/commands/ui-start.ts:27]:
```suggestion
import path from 'path';
import open from 'open';
```

File: js/core/src/reflection.ts [hunk 3]
```diff
@@ -227,6 +232,15 @@ export class ReflectionServer {
       response.json(this.options.configuredEnvs);
     });
 
+    server.post('/api/notify', async (request, response) => {
+      const { telemetryServerUrl } = request.body;
+      if (typeof telemetryServerUrl === 'string') {
+        setTelemetryServerUrl(telemetryServerUrl);
+        logger.info(`Connected to telemetry server on ${telemetryServerUrl}`);
```
Comment (pavelgj) [inline js/core/src/reflection.ts:239]: `logger.info` is forbidden inside the core framework

```suggestion
        logger.debug(`Connected to telemetry server on ${telemetryServerUrl}`);
```

File: js/core/src/reflection.ts [hunk 4]
```diff
@@ -245,11 +259,12 @@ export class ReflectionServer {
     });
 
     this.port = await this.findPort();
-    this.server = server.listen(this.port, () => {
-      logger.debug(
-        `Reflection server running on http://localhost:${this.port}`
+    this.server = server.listen(this.port, async () => {
+      logger.info(
```
Comment (pavelgj) [inline js/core/src/reflection.ts:263]:
```suggestion
      logger.debug(
```

File: js/core/src/tracing/exporter.ts [hunk 5]
```diff
@@ -130,6 +134,9 @@ export class TraceServerExporter implements SpanExporter {
   }
 
   private async save(traceId, spans: ReadableSpan[]): Promise<void> {
+    if (!telemetryServerUrl) {
+      return;
```
Comment (pavelgj) [inline js/core/src/tracing/exporter.ts:138]:
```suggestion
      logger.debug(`Telemetry server is not configured, trace ${traceId} not saved!`);
      return;
```

**Review by pavelgj — APPROVED** [review]

**Comment by pavelgj** [comment]
I tried it out, works great! Just a few minor fixes.
If you're still around please merge (if not, I'll just merge it myself next week).

**Comment by tonybaroneee** [comment]
@apascal07 how do we start up genkit itself now? what changes are needed in the sample apps to have `npm start` to start the reflect/flows/telemetry servers?

**Comment by pavelgj** [comment]
> @apascal07 how do we start up genkit itself now? what changes are needed in the sample apps to have `npm start` to start the reflect/flows/telemetry servers?

The script just needs to set `GENKIT_ENV=dev` env var...
We were thinking

```json
"scripts": {
  "genkit:dev": "GENKIT_ENV=dev npm run dev",
  "dev": "node --watch src/index.js",
}
```

the only problem is that setting env vars this way is not windows friendly. With some helpers:

```bash
npm i -D cross-env
```

```json
"scripts": {
  "genkit:dev": "cross-env GENKIT_ENV=dev npm run dev"
}
```

A full combo might be:


```bash
npm i -D cross-env concurrently
```

```json
"scripts": {
    "genkit:dev": "cross-env GENKIT_ENV=dev npm run dev",
    "dev": "concurrently \"npm build:watch\" \"node --watch lib/index.js\"",
    "build:watch": "tsc --watch",
}
```

or maybe we should lean into tsx more...

```bash
npm i -D cross-env tsx
```

```json
"scripts": {
    "genkit:dev": "cross-env GENKIT_ENV=dev npm run dev",
    "dev": "tsx --watch src/index.ts",
}
```

**Pull request merged**